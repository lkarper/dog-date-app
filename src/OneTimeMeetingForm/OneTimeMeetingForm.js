import React, { useState, useEffect } from 'react';
import OneTimeMeetingTimeWindow from '../OneTimeMeetingTimeWindow/OneTimeMeetingTimeWindow';

const OneTimeMeetingForm = (props) => {

    const { currentData, setOneTimeMeetingWindows } = props;
    
    const [date, setDate] = useState(currentData.date);
    const [timeWindows, setTimeWindows] = useState(currentData.timeWindows);

    useEffect(() => {

        if (currentData.date !== date) {
            setOneTimeMeetingWindows({
                date,
                timeWindows
            });
        }

        let update = false;

        if (timeWindows.length !== currentData.timeWindows.length) {
            update = true;
        }

        timeWindows.forEach(window => {
            if (!currentData.timeWindows.find(tw => tw.startTime === window.startTime && tw.endTime === window.endTime)) {
                update = true;
            }
        });

        if (update) {
            setOneTimeMeetingWindows({
                date,
                timeWindows
            });
        }

    }, [date, timeWindows, currentData, setOneTimeMeetingWindows]);

    const updateTimeWindows = (index, startTime, endTime) => {
        const updatedTimeWindow = {
            startTime,
            endTime,
        };

        const newTimeWindows = [...timeWindows];
        newTimeWindows.splice(index, 1, updatedTimeWindow);
        setTimeWindows(newTimeWindows);
    }

    const removeTimeWindow = (index) => {
        const updatedTimeWindows = [...timeWindows];
        updatedTimeWindows.splice(index, 1);
        setTimeWindows(updatedTimeWindows);
    }

    const addTimeWindow = () => {
        const newTimeWindow = {
            startTime: '',
            endTime: '',
        };

        const newTimeWindows = [...timeWindows, newTimeWindow];
        setTimeWindows(newTimeWindows);
    }

    const today = new Date();
    const minDate = `${today.getFullYear()}-${today.getMonth() + 1 < 10 ? `0${today.getMonth() + 1}` : today.getMonth() + 1}-${today.getDate() < 10 ? `0${today.getDate()}` : today.getDate()}`;

    return (
        <fieldset className="sub-fieldset">
            <legend>Looking for a one-time meeting on a specific date:</legend>
            <label htmlFor="once-date">Date:</label>
            <input 
                type="date" 
                id="once-date" 
                name="once-date" 
                placeholder="yyyy-mm-dd" 
                min={minDate}
                value={date}
                onChange={(e) => setDate(e.target.value)}
                aria-describedby='date-validator'
            />
            <ol>
                {timeWindows.map((window, i) => 
                    <OneTimeMeetingTimeWindow 
                        key={i}
                        index={i}
                        initialData={window}
                        updateTimeWindows={updateTimeWindows}
                        removeTimeWindow={removeTimeWindow}
                    />) 
                }
            </ol>
           <button type="button" onClick={addTimeWindow}>Add another time window on this date</button>
        </fieldset>
    );
}

export default OneTimeMeetingForm;