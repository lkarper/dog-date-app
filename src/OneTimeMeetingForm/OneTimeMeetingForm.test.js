import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import OneTimeMeetingForm from './OneTimeMeetingForm';

describe('OneTimeMeetingForm component', () => {
    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<OneTimeMeetingForm />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(<OneTimeMeetingForm />);
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });
});
