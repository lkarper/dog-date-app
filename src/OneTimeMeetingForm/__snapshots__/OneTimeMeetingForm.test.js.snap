// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneTimeMeetingForm component renders the UI as expected 1`] = `
<OneTimeMeetingForm
  currentData={
    Object {
      "date": "",
      "timeWindows": Array [
        Object {
          "endTime": "",
          "startTime": "",
        },
      ],
    }
  }
  setOneTimeMeetingWindows={[Function]}
>
  <fieldset
    className="sub-fieldset"
  >
    <legend>
      Looking for a one-time meeting on a specific date:
    </legend>
    <label
      htmlFor="once-date"
    >
      Date:
    </label>
    <input
      aria-describedby="date-validator"
      id="once-date"
      min="2020-09-16"
      name="once-date"
      onChange={[Function]}
      placeholder="yyyy-mm-dd"
      type="date"
      value=""
    />
    <ol
      aria-live="polite"
    >
      <OneTimeMeetingTimeWindow
        index={0}
        initialData={
          Object {
            "endTime": "",
            "startTime": "",
          }
        }
        key="0"
        removeTimeWindow={[Function]}
        type=""
        updateTimeWindows={[Function]}
      >
        <li
          className="OneTimeMeetingTimeWindow__li"
        >
          <div
            className="OneTimeMeetingTimeWindow__time-input-container"
          >
            <label
              htmlFor="once-start-time-1"
            >
              I'm available from: 
            </label>
            <input
              aria-describedby="start-time-validator-1"
              id="once-start-time-1"
              name="once-start-time-1"
              onChange={[Function]}
              placeholder="10:30"
              type="time"
              value=""
            />
          </div>
          <div
            className="OneTimeMeetingTimeWindow__time-input-container"
          >
            <label
              htmlFor="once-end-time-1"
            >
              I'm available until: 
            </label>
            <input
              aria-describedby="end-time-validator-1"
              id="once-end-time-1"
              min=""
              name="once-end-time-1"
              onChange={[Function]}
              placeholder="16:00"
              type="time"
              value=""
            />
          </div>
        </li>
      </OneTimeMeetingTimeWindow>
    </ol>
    <button
      className="OneTimeMeetingForm__button button"
      onClick={[Function]}
      type="button"
    >
      Add another time window on this date
    </button>
  </fieldset>
</OneTimeMeetingForm>
`;
